File: src/java/linkedlists/DoublyLinkedList.java
package java.linkedlists;

/**
 * Implement a doubly linked list 1 <-> 2 <-> 3.
 */

public class DoublyLinkedList {
    public static class Node {
        int val;
        Node next;
        Node prev;
        Node(int val) {
            this.val = val;
        }
    }

    private Node head;
    private Node tail;
    public DoublyLinkedList() {}

    public Node getHead() {
        return head;
    }

    public void addNode(int d) {
        Node node = new Node(d);
        if (head == null) {
            head = tail = node;
        } else {
            tail.next = node;
            node.prev = tail;
            tail = node;
        }
    }
}

File: src/main/kotlin/lists/DoublyLinkedList.kt
package lists

class DoublyLinkedList {
    class Node(val value: Int, var next: Node? = null, var prev: Node? = null)

    var head: Node? = null
    var tail: Node? = null

    fun add(v: Int) {
        val node = Node(v)
        if (head == null) {
            head = node
            tail = node
        } else {
            tail?.next = node
            node.prev = tail
            tail = node
        }
    }
}